### PR TITLE
Navigator state transfer [WIP]

### PR DESCRIPTION
**Pull request by adamgrzybowski**
# Goals
- Responsive view
- Fixed mounting/unmounting screens in the web navigator
- Navigation patterns described in the doc

## Responsive view
To achieve a responsive view (narrow or wide) we are switching between `NativeStackNavigator` and `WebNavigator` depending on the window width. Both are using `StackRouter` so the actions should transform the `react-navigation` state in the same way. The difference is in the way how navigators render the routes. We use a custom layout in the `WebNavigator`. 

When we switch between these two, we are using `react-native` state rehydration to generate new keys. There is a problem if we want to reuse the old keys for a different navigator. For some reason state in a nested stack is wiped out. There may be more problems that we haven't found. 

I guess that it has to do something with caching mechanism but it would need further investigation.

I am assuming that it is safer to get new keys instead of using the old ones but we can discuss that.

### `react-navigation` patch

If are navigating between screens on the web, the navigation state for a route is going to be "tied" to the history entry. If we are going back (or forward), `react-navigation` is going to check if there is a state tied to the revisited URL. If it exists, it is going to try to call the action `RESET` on the root navigator. 

**Problem:** 
When we go back to the URL that was opened with a different navigator. The `RESET` will fail because now the root navigator has a new key.

**Solution:** 
There is a function that strips keys and removes the `stale` property from the saved state and then uses the stripped state as an argument for the `RESET` action. In that case, `react-navigation` will generate a state with new keys.

## Fixed mounting/unmounting screens in the web navigator
Before this PR, the `WebNavigator` iterated on `routes` and conditionally rendered only visible screens.

**Problem:** (only for `WebNavigator`)
Let's look at the situation where the last route is a nested navigator and we new route for example "Search". In that case, nested navigator is no longer rendered because it shouldn't be visible. That causes the component to unmount. Because nested navigator is unmounted the state of this navigator is lost. The effect is inconsistent navigation behavior on the web/mobile.

**Solution:**
Instead of conditionally rendering we can conditionally change the style property `display:` to `none`. This way components aren't going to unmount but the browser is not going to render them.

This is how `NativeStackNavigator` deals with rendering routes.

## Navigation patterns described in the [doc](https://docs.google.com/document/d/1cQ87HcDA9RrmBM4ch7PmlNi9N9ZP8abQuPF0gKYrxDE/edit)
To achieve the expected behavior of nested navigators there is a helper function `navigateUp(<navigation>, <screenName>)`. In the POC example, it is used in the `About` screen for UP action (chevron left).

The `navigateUp` check if the desired screen is one route below the current route. 
If it is: `navigation.pop()` is called
If not: `navigation.replace(<screenName>)` is called

let's look at the example states:

**Opening the `about` directly**
```js
// UP action
['/r/1'] // - about > push(about)
['/r/1', ['about']] // - UP to the settings > replace('settings')
['/r/1', ['settings']]

// BACK action
['/r/1'] // - about > push(about)
['/r/1', ['about']] // - BACK > previous route for the browser, pop() for the android
['/r/1'] 
```

**Opening the `about` through the `settings`**
```js
// UP action
['/r/1'] // - settings >  push(settings)
['/r/1', ['settings']] // - about > push(about)
['/r/1', ['settings', 'about']] // - UP to the settings >  pop() 
['/r/1', ['settings']]

// BACK action
['/r/1'] // - settings >  push(settings)
['/r/1', ['settings']] // - about > push(about)
['/r/1', ['settings', 'about']] // - BACK  >  previous route for the browser, pop() for the android
['/r/1', ['settings']]
```

With this approach, the back browser behavior matches the back button behavior on the android.

## Review note
If you want to look at changes of `createWebViewNavigator.jsx` please use this [diff](https://github.com/marcaaron/navigation-test/pull/8/files#diff-aa0a9e400aad557fcabb65a6d6d81e34c1739364376de53c022678d1a970c091). I formatted this file with prettier so it would have the same indentation as the rest of files.

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import 'react-native-gesture-handler';
 import _ from 'underscore';
 import * as React from 'react';
-import {NavigationContainer, DefaultTheme, CommonActions, getStateFromPath, useFocusEffect} from '@react-navigation/native';
+import {NavigationContainer, DefaultTheme, CommonActions, getStateFromPath, useFocusEffect, useRoute, createNavigationContainerRef, getPathFromState } from '@react-navigation/native';
```
Comment (adamgrzybowski) [inline App.js]: Yep, I forgot to remove unused imports... I am going to fix this tomorrow 😄

File: App.js [hunk 2]
```diff
@@ -12,41 +12,53 @@ const chatTitleStyle = {fontSize: 18, fontWeight: 'bold'};
...
     Chat: 'r/:id?',
     LeftHandNav: '',
     SettingsStack: {
+      initialRouteName: 'Settings',
+      path: 'settings',
       screens: {
-        Settings: 'settings',
-        About: 'settings/about',
+        Settings: '',
+        About: 'about',
```
Comment (adamgrzybowski) [inline App.js]: With this setup, we don't have to add `LeftHandNav` and `settings` manually.

File: App.js [hunk 3]
```diff
@@ -11,42 +11,55 @@ const chevronStyle = {width: 30, height: 30, resizeMode: 'contain', marginRight:
...
+  const rawState = Array.isArray(state) ? [] : {};
+  for (const key of Object.keys(state)) {
+    if (keysBlocklist.includes(key)) {
+      continue
+    }
+    rawState[key] = typeof state[key] === 'object' ? stripNavigationState(state[key]) : state[key]
+  }
+  return rawState 
+}
+
```
Comment (adamgrzybowski) [inline App.js]: This function is used to remove `keys` and `stale` from the navigation state so we can regenerate it. Keeping the same `keys` for both navigators causes some problems. BTW not sure about `routeNames` it may be ok to leave them.

File: App.js [hunk 4]
```diff
@@ -11,42 +11,55 @@ const chevronStyle = {width: 30, height: 30, resizeMode: 'contain', marginRight:
...
   prefixes: ['http://localhost', 'https://navigation-test-lime.vercel.app'],
   config,
   getStateFromPath(path, cfg) {
     const state = getStateFromPath(path, cfg);
     console.log('Get state from path: ', state);
-
-    // We need to add the LHN if it does not exist
-    if (!_.find(state.routes, r => r.name === 'LeftHandNav')) {
-      state.routes.splice(0, 0, {name: 'LeftHandNav'});
-    }
```
Comment (adamgrzybowski) [inline App.js]: The new config handles this.

File: App.js [hunk 5]
```diff
@@ -231,51 +253,46 @@ const navTheme = {
...
-
-      // Since About is in a Settings stack we want the "up" button to go back to the Settings main page and so need to define this
-      if (pathname === '/settings/about') {
-        state.routes[1].state.routes.unshift({name: 'Settings'});
-      }
-
-      // Add a report when we are on large format web and a chat does not exist in the route
-      if (!isSmallScreenWidth && !state.routes.find(route => route.name === 'Chat')) {
-          state.routes.splice(1, 0, {name: 'Chat', params: {id: 1}});
-      }
```
Comment (adamgrzybowski) [inline App.js:267]: I am not sure if the `initialState` is necessary? It looks like we can handle this just with the right linking config

**Comment by adamgrzybowski** [comment]
Yeap, I overlooked one of the most basic use cases 🤦  (widescreen, mobile device) and `initialState` may be actually needed. I will fix it tomorrow.

File: App.js (hunk 3 above)
Comment (marcaaron) [reply App.js]: `routeNames` seems like it wouldn't change. But it also seems like something we wouldn't need to include in an `initialState` or state reset.

Comment (marcaaron) [reply App.js]: I'm not too sure.

Comment (marcaaron) [reply App.js]: I'm also a little unsure what the "stale" things in the state are used for (even after reading [this](https://reactnavigation.org/docs/navigation-state/#partial-state-objects)) but it seems like something we might not need to worry about.

File: App.js (hunk 5 above)
Comment (marcaaron) [reply App.js:267]: I'm not sure if the config can be used to do everything (but maybe there's a way to also make it aware of the screen dimensions).

One breaking change here is that if we navigate directly to the site root on mobile web we are directed to a chat and the expected behavior would be to remain on the sidebar.

File: App.js (hunk 3 above)
Comment (techievivek) [reply App.js]: To be extra safe it makes sense to get rid of the `stale` key or set it to `true` otherwise react-navigation will not fix any issue in the state object.

**From the doc:**

If you want React Navigation to fix invalid state, you need to make sure that you don't have `stale: false` in the state object. State objects with `stale: false` are assumed to be valid state objects and React Navigation won't attempt to fix them.

---------------------------------------------------------------------------
Also based on the above assumption that react-navigation will handle fixing state for us it should be fine to not include `routeNames`.

**From the doc:**
you can only provide a routes array without any keys and React Navigation will automatically add everything that's needed to make it work:
```js
const state = {
  routes: [{ name: 'Home' }, { name: 'Profile' }],
};
```
After rehydration, it'll look something like this:
```js
const state = {
  type: 'stack',
  key: 'stack-1',
  routeNames: ['Home', 'Profile', 'Settings'],
  routes: [
    { key: 'home-1', name: 'Home' },
    { key: 'settings-1', name: 'Settings' },
  ],
  index: 1,
  stale: false,
};
```
Here, React Navigation filled in the missing bits such as keys, route names, index etc.

File: App.js [hunk 6]
```diff
@@ -11,42 +11,55 @@ const chevronStyle = {width: 30, height: 30, resizeMode: 'contain', marginRight:
...
+        About: 'about',
       },
     },
     Search: 'search'
   },
 };
 
+
+const addChatRouteIfNecessary = (state) => {
+    if (!_.find(state.routes, r => r.name === 'Chat')) {
```
Comment (techievivek) [inline App.js]: Based on what Marc pointed out [here](https://github.com/marcaaron/navigation-test/pull/8/files#r1054838939) I think we can conditionally add a `chat` screen.
```js
if(path === '/' && !checkIsSmallScreen(Dimensions.get('window').width)) {
    addChatRouteIfNecessary(state)
}
```

**Comment by mountiny** [comment]
Agreed with the other comments, lets see how we can improve this.

File: App.js (hunk 6 above)
Comment (adamgrzybowski) [reply App.js]: Good idea! I've used it in the fixes. The only difference is that I've put it in the `addChatRouteIfNecessary` BTW If you have a better name for this function please share 🙏  For some reason I don't like it but can't think of anything better.

Comment (marcaaron) [reply App.js]: maybe because it is exposing an implementation details that we are "adding a chat route" and also it "might not happen". So, a more general name like `fixState()` could be OK and if you need to know what that means you go look at the function.

File: App.js (hunk 3 above)
Comment (marcaaron) [reply App.js]: Thanks @techievivek, good notes on the rehydrated state. I think that explains well why we do not need `routeNames` or `key`.

Next, I agree that we could remove the `stale` property. But my point above is that I am not sure whether we **need to**. 

If the `state` is "stale" then it sounds like we might not want to use it at all? I'd like to understand more about whether the `state` will ever be missing `stale: false` when passing through this function before we make an effort to remove it. Otherwise, we are just doing stuff without a clear picture of why it matters.

Comment (marcaaron) [reply App.js]: Like this...

>If you want React Navigation to fix invalid state, you need to make sure that you don't have stale: false in the state object

When do we want React Navigation to fix the "invalid state" and what does it mean? Do we want it to do this or not?

File: App.js [hunk 7]
```diff
@@ -1,52 +1,70 @@
...
-  prefixes: ['http://localhost', 'https://navigation-test-lime.vercel.app'],
-  config,
-  getStateFromPath(path, cfg) {
-    const state = getStateFromPath(path, cfg);
-    console.log('Get state from path: ', state);
 
-    // We need to add the LHN if it does not exist
-    if (!_.find(state.routes, r => r.name === 'LeftHandNav')) {
-      state.routes.splice(0, 0, {name: 'LeftHandNav'});
+const addChatRouteIfNecessary = (state) => {
```
Comment (mountiny) [inline App.js]: Can we update this to 

```suggestion
const fixState = (state) => {
```

**Review by mountiny — COMMENTED** [review]
@adamgrzybowski iOS working great

One issue I run into on web which is probably related to our discussions in slack:
1. Be on the main page
2. Go to Settings
3. Go to About
4. Click the Search link in the about page
5. Click the link to first chat from the Search page
6. As you are on the first chat page - click browser back
7. This should take you to the Search page, but it does nothing
8. Second Back click brings you to Settings/About

So it skipped over the search page basically

I think it would be great to polish this PR up and then merge this once we will be happy with it.

I will start the discussion today in the open source channel and we can then reaffirm we are all fine with how the back button in browser should behave

File: App.js [hunk 8]
```diff
@@ -42,12 +45,12 @@ const config = {
...
 
 
-const addChatRouteIfNecessary = (state) => {
+const fixState = (state) => {
     // we don't want to add Chat route for small screens
     if (checkIsSmallScreen(Dimensions.get('window').width)) {
       return 
     }
-    if (!_.find(state.routes, r => r.name === 'Chat')) {
+    if (!find(state.routes, r => r.name === 'Chat')) {
```
Comment (mountiny) [inline App.js]: I know this is just POC, but I would keep the underscore here, we use that in the App

**Review by mountiny — COMMENTED** [review]
Testing really well! Lets look into cleaning this up _a little bit_ so we can merge this unless there are some issue @techievivek would you like to stress test this with our high-level design as well?

**Comment by techievivek** [comment]
@mountiny Yes, I did have a look at it yesterday but changes were not pushed then so I will give it one more look today.

File: patches/@react-navigation+native+6.0.13.patch [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+     const onStateChange = async () => {
+       const navigation = ref.current;
+-
+       if (!navigation || !enabled) {
+         return;
+       }
+-
+       const previousState = previousStateRef.current;
+-      const state = navigation.getRootState(); // root state may not available, for example when root navigators switch inside the container
++      const state = navigation.getRootState();
```
Comment (mountiny) [inline patches/@react-navigation+native+6.0.13.patch]: @adamgrzybowski could we clean up this patch? I see there is lots of the changes just styling or removing comments, which then makes it hard to follow what are the actually important changes.

Additionally, let's ass some good explanation comments to any change we had to do in this patch, I know this patch is mainly to endsure the state is not broken after resizing from mobile to desktop layout, but it woul dbe great to have only required changes in the diff here.

File: App.js [hunk 10]
```diff
@@ -64,23 +83,29 @@ function WithCustomBackBehavior(props) {
...
+  return state.routes?.[state.index]?.params
+}
+
 function LeftHandNav({navigation}) {
+  const focusedRouteParams = useFocusedRouteParams(navigation);
   return (
     <View style={{flex: 1}}>
       <LHNHeader navigation={navigation} />
-      <Pressable style={{flexDirection: 'row', margin: 10, alignItems: 'center'}} onPress={() => navigation.push('Chat', {id: 1})}>
+      <Pressable style={{flexDirection: 'row', margin: 10, alignItems: 'center'}} disabled={focusedRouteParams?.id === 1} onPress={() => navigation.push('Chat', {id: 1})}>
```
Comment (techievivek) [inline App.js:79]: Thanks for fixing this up. Any idea why `route` doesn't contain the `params` here?

File: App.js [hunk 11]
```diff
@@ -169,7 +194,7 @@ function SettingsScreen({navigation}) {
         </Pressable>
         <Text style={titleStyle}>Settings</Text>
       </View>
-      <Pressable style={{flexDirection: 'row', marginTop: 10}} onPress={() => navigation.push('About')}>
+      <Pressable style={{flexDirection: 'row', marginTop: 10}} onPress={() => navigation.replace('About')}>
```
Comment (techievivek) [inline App.js]: Using `replace` here seems to break the navigation(as per the doc) if the user navigates from LHNHeader to `about` page.

Navigation: `LHNHeader -> Settings -> About`.

Actual:
- Pressing `back` when the user is on `about` page takes user back to chat screen.

Expected:
- Pressing `back` when the user is on `about` page takes user back to `settings` screen.

https://user-images.githubusercontent.com/35863227/214007098-5be81065-bd14-4374-bfbb-7e92cb55e84c.mov

Comment (techievivek) [reply App.js]: Why can't we keep `push` here? It doesn't seem to be breaking anything, right? and also, it makes sense because you are a pushing a new screen in the current stack.

Comment (mountiny) [reply App.js]: ~I think that this behaviour is expected per the design doc~

Discussing in DM, its not very clear from high-level design what should be done here but based on some previous discussions we should indeed go "up" as Vivek mentions.

File: App.js (hunk 10 above)
Comment (adamgrzybowski) [reply App.js:79]: It does but it is a route for the LHN. We need to get top chat params

File: App.js (hunk 11 above)
Comment (techievivek) [reply App.js]: Aah, I see why you added `replace` here instead of `push`(because it seems like we have to double tap the `up` button on setting screen to hide the modal).

Comment (techievivek) [reply App.js]: @adamgrzybowski Quick question: When you use `replace` is it meant to replace the whole nested route or just the current route? Because having `replace('Settings')` in `about` page just seems to remove the `about` route and add a new route for `settings`.

File: App.js (hunk 10 above)
Comment (techievivek) [reply App.js:79]: Ohhh I see that’s why it was undefined. Make sense to get the top chat’s param here.

File: patches/@react-navigation+native+6.0.13.patch [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
++        // The top level key of current state and cached state may be different.
++        // That will happen if the user switched between wide and narrow view,
++        // because we are using different navigators for these views.
++        // In that case we need to regenerate the state with new keys to make navigation work 
++        if (rootState.key !== record.state.key) {
++          navigation.resetRoot(stripKeysFromNavigationState(record.state));
++        } else {
++          navigation.resetRoot(record.state);
++        } 
+         return;
```
Comment (adamgrzybowski) [inline patches/@react-navigation+native+6.0.13.patch]: We want to do this only if necessary

File: App.js [hunk 13]
```diff
@@ -1,86 +1,82 @@
...
- * except the root screen of that nested navigator will need to implement this.
- */
-function WithCustomBackBehavior(props) {
-  useFocusEffect(() => {
-    const subscription = BackHandler.addEventListener('hardwareBackPress', () => {
-      props.navigation.popToTop();
-    });
-
-    return () => subscription.remove();
-  });
```
Comment (adamgrzybowski) [inline App.js]: It looks like we don't need this anymore

**Comment by techievivek** [comment]
@adamgrzybowski I am seeing an inconsistency in the below navigation patterns.
1. `LHNHeader -> Settings -> About -> Search` press `up` ends up in the `Settings` page.
2. `Chat -> About -> Search` press `up` ends up in the `About` page.


https://user-images.githubusercontent.com/35863227/214321804-189e4d30-8467-4677-8a69-1070263a6476.mov

Shouldn't it behave the same way when we press `up` on the search page?

P.S. This seems to be broken only on the web(Android seems to work just fine).

File: App.js (hunk 13 above)
Comment (techievivek) [reply App.js]: Yeah, the app seems to work properly for Android.

**Comment by adamgrzybowski** [comment]
EDIT: Solved

> @adamgrzybowski I am seeing an inconsistency in the below navigation patterns.
>
> LHNHeader -> Settings -> About -> Search press up ends up in the Settings page.
> Chat -> About -> Search press up ends up in the About page.

Yeah for some reason state for the `SettingsStack` is cleaned after pushing `Search`. In both cases. The difference Is for one method we are pushing `SettingsStack` with params `screen: 'About'` and for the other not. The one with params is opening given screen when going back.
image
image

Not sure why states disappear though

File: patches/@react-navigation+native+6.0.13.patch [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+diff --git a/node_modules/@react-navigation/native/lib/module/stripKeysFromNavigationState.js b/node_modules/@react-navigation/native/lib/module/stripKeysFromNavigationState.js
+new file mode 100644
+index 0000000..31ba628
+--- /dev/null
++++ b/node_modules/@react-navigation/native/lib/module/stripKeysFromNavigationState.js
+@@ -0,0 +1,17 @@
++// This function removes keys and some other properties from the navigation state 
++// in order to force react-navigation to regenerate those keys
++
++// keys - if we try to use the same keys it is going to break the navigation
```
Comment (marcaaron) [inline patches/@react-navigation+native+6.0.13.patch]: I feel like the natural question here is: Why does the navigation break? And what exactly breaks about it?

File: patches/@react-navigation+native+6.0.13.patch [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
++// stale - with stale: false navigation won't regenerate the state 
++// routeNames - to be sure that we have updated routeNames if something changed 
++
++// The argument for this function may be: NavigationState, NavigatoinRoute, PartialState<NavigationState>
```
Comment (marcaaron) [inline patches/@react-navigation+native+6.0.13.patch]:
```suggestion
+// The argument for this function may be: NavigationState, NavigationRoute, PartialState<NavigationState>
```

File: patches/@react-navigation+native+6.0.13.patch [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ import createMemoryHistory from './createMemoryHistory';
+@@ -171,7 +172,16 @@ export default function useLinking(ref, _ref) {
+       const record = history.get(index);
+ 
+       if ((record === null || record === void 0 ? void 0 : record.path) === path && record !== null && record !== void 0 && record.state) {
+-        navigation.resetRoot(record.state);
++        const rootState = navigation.getRootState();
++        // The top level key of current state and cached state may be different.
++        // That will happen if the user switched between wide and narrow view,
++        // because we are using different navigators for these views.
```
Comment (marcaaron) [inline patches/@react-navigation+native+6.0.13.patch]: Just rubber ducking here to confirm that I get the patch changes...

- It looks like this code is running when the user pops a screen in the browser.
- A "record" in this case is like an entry in the memory history `items` array which has a picture of the state we should be going back to. 
- And it looks like we are comparing the `rootState.key` (react navigations internal state) with the "record" state key to see if they are the same.
    - This is used as a signal to tell that we switched to a different navigator.

So is the only edge case we are handling here something like this...

- User is on web + a wide screen view
- User switches to a narrow screen view
- User tries to go back
- 💥 because the internal state key that react navigation has does not match what is in the memory history `items` array?

?

File: patches/@react-navigation+native+6.0.13.patch [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
++        // In that case we need to regenerate the state with new keys to make navigation work
```
Comment (marcaaron) [inline patches/@react-navigation+native+6.0.13.patch]: And I guess a follow up question would be...

Is it bad to propose that `react-navigation` _should_ support switching from one navigator to another and that the memory history should not necessarily need to have the exact same keys as the navigator?

I think I'm not quite understanding why the memory history's default behavior would be to break unexpectedly if these keys are mismatching. And if there's not some more formal route we could take to simply updating the memory history entries...

I don't know if either of these make sense to explore, but I was curious if either of them could work...

**Idea 1**

- We explicitly detect the navigator switch
- The memory history is rebuilt with new state keys

**Idea 2**

- We explicitly detect the navigator switch
- The keys are preserved instead of regenerated

**Review by marcaaron — COMMENTED** [review]
Changes are looking good. I left a few questions about the patch. Mostly trying to get a really good idea about why we need it since I think it's still a little unclear for me.

File: patches/@react-navigation+native+6.0.13.patch (hunk 14 above)
Comment (adamgrzybowski) [reply patches/@react-navigation+native+6.0.13.patch]: Here is an example of what is happening if we don't remove keys. Unwanted jump from the settings screen to the about screen when resizing. It just works if we remove keys before.

https://user-images.githubusercontent.com/67908363/214559461-bd32365d-b0f7-40c2-aad1-c5ad0059c60b.mov

It looks like for some reason nested state is wiped out. I am not sure what exactly is happening under the hood but I could look for the exact reason if you want. But I am afraid that in the end we will be forced to get new keys anyway. 

It kind of makes sense to me to use different keys for different navigators. Those navigators give us a completely different structure of components and the keys are used as the keys for these components.

File: patches/@react-navigation+native+6.0.13.patch (hunk 17 above)
Comment (adamgrzybowski) [reply patches/@react-navigation+native+6.0.13.patch]: What we are doing here is conditionally switching the root navigator. How the error looks exactly is 
image

So the root navigator has a different `key` and therefore the `RESET` action is not handled properly. 

My guess why `react-navigation` isn't supporting this use case is the fact that it's probably very rare to change the root navigator like that and use the same state for two different navigators.

**Idea 1**
The problem is the fact that history entries may and will have a different "shape" than our current navigation state. So it wouldn't be one to one to replace keys in the history entries with newly generated keys

**Idea 2**
I think that the same problem may exist here. (If I understand the idea correctly)

File: patches/@react-navigation+native+6.0.13.patch (hunk 16 above)
Comment (adamgrzybowski) [reply patches/@react-navigation+native+6.0.13.patch]: That's correct. I posted a screenshot of the error in my response below. Take a look

**Review by mountiny — COMMENTED** [review]
Tested the POC today and worked well! Resizing from desktop to mobile and back worked well too as well as native. 

Thanks for writing the detailed description, I think it looks very good and makes sense.

File: patches/@react-navigation+native+6.0.13.patch (hunk 14 above)
Comment (marcaaron) [reply patches/@react-navigation+native+6.0.13.patch]: >It looks like for some reason nested state is wiped out. I am not sure what exactly is happening under the hood but I could look for the exact reason if you want.

Yeah, I'd love to know the answer. It's great that we have some solution. But I'm having trouble understanding what the problem is exactly.

>It kind of makes sense to me to use different keys for different navigators. Those navigators give us a completely different structure of components and the keys are used as the keys for these components.

What makes sense about it?

**quick aside:** I am asking these questions because I anticipate that other stakeholders will ask them. The more robust an explanation we have the easier it is for us to get buy in on stuff or think of alternate solutions. Personally, I am really happy with the progress overall.

File: patches/@react-navigation+native+6.0.13.patch (hunk 17 above)
Comment (marcaaron) [reply patches/@react-navigation+native+6.0.13.patch]: >it's probably very rare to change the root navigator like that and use the same state for two different navigators

I hear you that this is a more advanced use-case. But, I'm still curious about this one...

>Is it bad to propose that react-navigation should support switching from one navigator to another and that the memory history should not necessarily need to have the exact same keys as the navigator?

Maybe there's some reasonable argument we can make that this _should_ be supported?

>The problem is the fact that history entries may and will have a different "shape" than our current navigation state. So it wouldn't be one to one to replace keys in the history entries with newly generated keys

Ah sorry, I'm not sure if I understood this. I'm sure I'm just missing something obvious here. Can we discuss an example where "history entries may and will have a different "shape" than our current navigation state"?

**Comment by marcaaron** [comment]
Agree the description is really clear, thanks for updating this! 😍 

>For some reason state in a nested stack is wiped out. There may be more problems that we haven't found.
>I guess that it has to do something with caching mechanism but it would need further investigation.
>I am assuming that it is safer to get new keys instead of using the old ones but we can discuss that.

Already left some review comments about this - but I think having a clear picture of the problem will help people understand the solution better. We have a process where the entire company reviews the proposal so it needs to be rock solid.

The `display: none` and `navigateUp()` solutions both look very reasonable to me. I think people will struggle the most with the patch - and will be likely to ask whether there's any chance to prioritize an upstream change.

File: patches/@react-navigation+native+6.0.13.patch (hunk 17 above)
Comment (adamgrzybowski) [reply patches/@react-navigation+native+6.0.13.patch]: To clarify, idea nr 1 looks like this:
- navigate a few times somewhere
- resize window
- new keys generated for the current navigation state so for example: key for the root navigator was `abc` now is `xyz`
- we modify every item in the memory history so the root key is now `xyz` in every saved state



Is that correct?

Comment (marcaaron) [reply patches/@react-navigation+native+6.0.13.patch]: Yes! Basically, the memory history is in "memory" so I'm wondering if we can modify or update it directly so that it references the newly generated keys for the new navigator. That way, when we go back or forward - the memory history keys match the navigation state already and no changes in `useLinking` need to happen.

Idea nr 2 is more like...

Why must the keys change at all if we switch to a new root navigator? I am sure there are some general gaps in my knowledge about what `react-navigation` uses the keys for that might explain why this is difficult to do. And it could also be that nobody ever thought about it because 99% of the apps will have one root navigator that never changes.

Comment (adamgrzybowski) [reply patches/@react-navigation+native+6.0.13.patch]: I don't think we have direct access to history memory from the app level. It would need a patch anyway if we want to change states saved there

And about the "shape"

let's assume a situation like that:
```
/r/1                  -> ['chat']
/settings/about       -> ['chat', ['about]]
/search               -> ['chat', ['about'], 'search']        // go back with the browser back button
/settings/about       -> ['chat',['about]]                    // now resize
```

This is how history memory looks like now:
```
/r/1                   -> ['chat']                               // 0
/settings/about        -> ['chat', ['about]]                     // 1 <- browser history points here 
/search                -> ['chat', ['about'], 'search']          // 2
```


Our navigation state is `['chat', ['about]]`  and we are generating new keys for these routes so we have keys for `chat` and `about`.

Now we need to update all saved states in history memory
```
['chat']
['chat', ['about]]
['chat', ['about'], 'search']
```

What should be the key for the search route? We don't have it because the route wasn't present in the navigation state when we resized the window

And this state `['chat', ['about'], 'search']` would be used if we press the forward browser button

Comment (adamgrzybowski) [reply patches/@react-navigation+native+6.0.13.patch]: https://user-images.githubusercontent.com/67908363/214663854-8ccd715f-c6b5-44b8-a75b-eddf454b61cc.mov

Here's how it would look like in practice

Comment (marcaaron) [reply patches/@react-navigation+native+6.0.13.patch]: >I don't think we have direct access to history memory from the app level. It would need a patch anyway if we want to change states saved there

Correct. I agree there could be alternative solutions. 👍 

>What should be the key for the search route? We don't have it because the route wasn't present in the navigation state when we resized the window

I think I'm missing something about what the state looks like after go "back" and pop a screen. 

When we go "back" in the example does the navigation state lose that screen (and it's key)? Or is it still retained somewhere in the navigation state? Assuming we never switch navigators, it seems like somehow the "forward" action still works correctly even though that screen is not in the navigation hierarchy anymore?

It previously sounded like the main issue is with the key of the root navigator? But now sounds like there is some problem with popped screens like the "Search" screen in our example?

Main question I am looking to get answered is something like...

What does `react-navigation` use the keys in the memory history for when reseting the root state? Do we need them at all?

Comment (techievivek) [reply patches/@react-navigation+native+6.0.13.patch]: > What does react-navigation use the keys in the memory history for when reseting the root state? Do we need them at all?

This one would definitely be worth looking into it. And great conversation so far.

File: App.js [hunk 18]
```diff
@@ -231,51 +249,94 @@ const navTheme = {
...
+    Linking.getInitialURL().then(this.handleInitialState)
   }
 
   render() {
-    if (!this.state.initialState) {
-      return null;
-    }
-
     return (
+      <IsSmallScreenContext.Provider value={this.state.isSmall}>
```
Comment (techievivek) [inline App.js]:
```suggestion
      <IsSmallScreenContext.Provider value={this.state.isSmallScreen}>
```

Comment (adamgrzybowski) [reply App.js]: Good catch! Thanks!

File: App.js [hunk 19]
```diff
@@ -231,51 +249,94 @@ const navTheme = {
...
+
+const WebStack = createWebNavigator();
+const NativeStack = createNativeStackNavigator();
+
+const SettingsStack = createNativeStackNavigator();
+
+const WebNavigator = createMainStack(WebStack);
+const NativeNavigator= createMainStack(NativeStack);
+
+const RootStack = createNativeStackNavigator();
```
Comment (techievivek) [inline App.js:293]: Hi, @adamgrzybowski. Can you share your thought process on why we are adding an extra layer of a navigator? I am having a bit of a hard time understanding this change, so if you can give some context that will be cool.

Comment (mountiny) [reply App.js:293]: In high-level it seems like this allows us to not use the patch (its gone now). But we definitely still need to get down to the keys problem (and why would be solved with the wrapping navigator)

**Comment by techievivek** [comment]
Also, Posting the below screen share for a quick clarification on what is the expected behaviour.


<video src="https://user-images.githubusercontent.com/35863227/215110601-201db20b-e479-490b-92f8-29c712d69073.mov"></video>

Pattern: `LHN -> Settings -> About -> Search -> Chat` clicks `up`

I would expect the user to land in `LHN`, right? If that's the case we have to consider this case in the `navigationUp` method as well.

File: App.js (hunk 19 above)
Comment (adamgrzybowski) [reply App.js:293]: Sure! 
1. The conditionally rendered navigators aren't really a pattern in `react-navigation` unless they are wrapped in a screen. With this approach, we are closer to the solutions presented in the react-navigation docs. That's also something my colleague pointed me. Being closer to the general use case we are less prone to bugs.
2. If we are conditionally rendering screens and not navigators that means we keep the same root navigator. The key for the root stays the same after resizing and we don't need to patch `react-navigation` to make it work when navigating back with the browser button. 
3. It will be easier to separate the logic for resizing. I haven't done it here yet but I guess that may be useful in the main app. Instead of `RootNavigator` we could call it `ResponsiveNavigator` or something like that.

**Comment by adamgrzybowski** [comment]
> I would expect the user to land in LHN, right? If that's the case we have to consider this case in the navigationUp method as well.

So we want clear the stack after opening a chat? @mountiny Can you confirm that this is a pattern that we want?

File: App.js (hunk 19 above)
Comment (techievivek) [reply App.js:293]: Looks promising, waiting for point 3 to be implemented. Going to give it some more testing for now.
